pkg/storage/chunkfs/provider.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2024 The Solaris Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package chunkfs
+
+import (
+	"context"
+	"fmt"
+	"github.com/solarisdb/solaris/golibs/container/iterable"
+	"github.com/solarisdb/solaris/golibs/errors"
+	"github.com/solarisdb/solaris/golibs/files"
+	"github.com/solarisdb/solaris/golibs/logging"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+)
+
+// Provider manages a pull of opened chunks and allows to return a Chunk object by request.
+// The Provider limits the number of opened file descriptors and the space on the local drive
+// borrowed for the chunks
+type Provider struct {
+	lock   sync.Mutex
+	cond   *sync.Cond
+	logger logging.Logger
+	dir    string
+	chunks map[string]*Chunk
+	active int
+	maxFDs int
+	// standBy is the collection of opened Chunks that maybe removed
+	standBy *iterable.Map[string, *Chunk]
+	// otOpenList the list of chunks that have to be opened
+	toOpenList []*Chunk
+	closed     bool
+	closedCh   chan struct{}
+}
+
+// NewProvider creates the new Provider instance
+func NewProvider(dir string, maxOpenedChunks int) *Provider {
+	if maxOpenedChunks <= 0 {
+		panic(fmt.Sprintf("expecting maxOpenedChunks=%d be positive", maxOpenedChunks))
+	}
+	p := new(Provider)
+	p.cond = sync.NewCond(&p.lock)
+	p.chunks = make(map[string]*Chunk)
+	p.logger = logging.NewLogger("chunkfs.Provider")
+	p.dir = dir
+	p.maxFDs = maxOpenedChunks
+	p.standBy = iterable.NewMap[string, *Chunk]()
+	p.closedCh = make(chan struct{})
+	go p.opener()
+	return p
+}
+
+// GetOpenedChunk returns an opened chunk (ready to be used) by its ID.
+// The function may return ctx.Err() or ErrClosed errors
+func (p *Provider) GetOpenedChunk(ctx context.Context, id string) (*Chunk, error) {
+	c, err := p.getOrCreateChunk(id)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case <-ctx.Done():
+		p.ReleaseChunk(c)
+		return nil, ctx.Err()
+	case <-p.closedCh:
+		p.ReleaseChunk(c)
+		return nil, errors.ErrClosed
+	case <-c.neverOpened:
+	}
+	return c, nil
```

Comment:
q: can we get a closed chunk? I mean, the opener hasn't been yet able to `c.Open()` it but we already tried to `c.Append`...

Reply:
it was not an idea, the closed chunk cannot be returned by the function. `c.neverOpened` is not closed before the chunk is opened.